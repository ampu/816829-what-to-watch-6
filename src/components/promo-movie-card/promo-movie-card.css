.movie-card--empty {
  padding-bottom: 0;
}

.movie-card__head--empty {
  margin-bottom: 0;
}
