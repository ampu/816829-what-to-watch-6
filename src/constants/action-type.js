const ActionType = {
  SET_MOVIES_STATUS: `movies/setMoviesStatus`,
  SET_MOVIES: `movies/setMovies`,
  SET_PROMO_STATUS: `movies/setPromoStatus`,
  SET_PROMO_MOVIE: `movies/setPromoMovie`,

  SET_LOGIN_STATUS: `user/setLoginStatus`,
  SET_USER: `user/setUser`,
  SET_MY_LIST_STATUS: `user/setMyListStatus`,
  SET_MY_LIST: `user/setMyList`,

  SET_GENRE: `genres/setGenre`,
};

export default ActionType;
