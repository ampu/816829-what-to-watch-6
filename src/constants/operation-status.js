const OperationStatus = {
  UNSET: `unset`,
  PENDING: `pending`,
  RESOLVED: `resolved`,
  REJECTED: `rejected`,
  UNAUTHORIZED: `unauthorized`,
};

export default OperationStatus;
