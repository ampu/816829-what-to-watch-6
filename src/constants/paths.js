const MainPath = {
  INDEX: `/`,
  SIGN_IN: `/login`,
  MY_LIST: `/mylist`,
  MOVIE: `/films/:id`,
  ADD_REVIEW: `/films/:id/review`,
  PLAYER: `/player/:id`,
  NOT_FOUND: `/not-found`,
};

const MoviePath = {
  MOVIE_OVERVIEW: `/films/:id`,
  MOVIE_DETAILS: `/films/:id/details`,
  MOVIE_REVIEWS: `/films/:id/reviews`,
};

const MOVIE_PATHS = Array.from(new Set([MainPath.MOVIE, ...Object.values(MoviePath)]));

export {
  MainPath,
  MoviePath,
  MOVIE_PATHS,
};
