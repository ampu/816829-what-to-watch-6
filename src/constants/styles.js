const DEFAULT_BACKGROUND_STYLE = {backgroundColor: `black`};
const BORDER_BOTTOM_STYLE = {borderBottomStyle: `rgba(255, 255, 255, 0.24)`};

export {
  DEFAULT_BACKGROUND_STYLE,
  BORDER_BOTTOM_STYLE,
};
