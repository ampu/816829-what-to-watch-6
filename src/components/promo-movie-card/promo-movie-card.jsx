import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import getClassName from 'classnames';

import OperationStatus from '../../constants/operation-status';
import {DEFAULT_BACKGROUND_STYLE} from '../../constants/styles';
import {selectPromoMovie, selectPromoStatus} from '../../store/selectors/movie-selectors';
import {getPromoMovie} from '../../store/operations/movie-operations';

import Logo from '../logo/logo';
import UserBlock from '../user-block/user-block';
import MoviePoster from '../movie-poster/movie-poster';
import MovieDescription from '../movie-description/movie-description';
import MovieBackground from '../movie-background/movie-background';
import SpinnerLoading from '../spinner-loading/spinner-loading';
import Container from '../container/container';

import {movieShape} from '../../typings/movie-type';
import operationStatusType from '../../typings/operation-status-type';

import './promo-movie-card.css';

const PromoMovieCard = ({onMounted, movieStatus, movie = {}}) => {
  const {
    title = ``,
    primaryBackgroundStyle = DEFAULT_BACKGROUND_STYLE,
  } = movie;

  useEffect(() => {
    onMounted();
  }, [onMounted]);

  const isEmpty = !movie.id;

  const classMap = {
    [`movie-card`]: true,
    [`movie-card--empty`]: isEmpty,
  };

  const headerClassMap = {
    [`page-header`]: true,
    [`movie-card__head`]: true,
    [`movie-card__head--empty`]: isEmpty,
  };

  return (
    <section className={getClassName(classMap)} style={primaryBackgroundStyle}>
      <MovieBackground movie={movie}/>

      <h1 className="visually-hidden">{title}</h1>

      <header className={getClassName(headerClassMap)}>
        <Logo/>

        {movieStatus === OperationStatus.PENDING && (
          <Container isCentered isAbsolute>
            <SpinnerLoading/>
          </Container>
        )}

        <UserBlock/>
      </header>

      <div className="movie-card__wrap">
        <div className="movie-card__info">
          <MoviePoster movie={movie}/>

          {movie.id && <MovieDescription movie={movie}/>}
        </div>
      </div>
    </section>
  );
};

PromoMovieCard.propTypes = {
  onMounted: PropTypes.func.isRequired,
  movieStatus: operationStatusType,
  movie: movieShape,
};

const mapStateToProps = (state) => ({
  movieStatus: selectPromoStatus(state),
  movie: selectPromoMovie(state),
});

const mapDispatchToProps = (dispatch) => ({
  onMounted() {
    return dispatch(getPromoMovie());
  },
});

export {PromoMovieCard};
export default connect(mapStateToProps, mapDispatchToProps)(PromoMovieCard);
