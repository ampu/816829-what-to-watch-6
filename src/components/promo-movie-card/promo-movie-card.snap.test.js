import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import * as redux from 'react-redux';

import {render} from '@testing-library/react';
import configureStore from 'redux-mock-store';

import OperationStatus from '../../constants/operation-status';
import {ALL_GENRES} from '../../constants/genre';
import Domain from '../../store/domain';

import {PromoMovieCard} from './promo-movie-card';

const INITIAL_STATE = {
  [Domain.MOVIE]: {
    moviesStatus: OperationStatus.UNSET,
    movies: [],
    promoStatus: OperationStatus.UNSET,
    promoMovie: undefined,
  },
  [Domain.USER]: {
    loginStatus: OperationStatus.UNSET,
    user: undefined,
  },
  [Domain.GENRE]: {
    genre: ALL_GENRES,
  },
};

const thunkMock = () => (next) => (action) => {
  if (typeof action === `function`) {
    return Promise.resolve();
  }
  return next(action);
};

const initializeStoreWithState = configureStore([thunkMock]);

const renderPromoMovie = (props = {}) => {
  return render((
    <redux.Provider store={initializeStoreWithState(INITIAL_STATE)}>
      <Router history={createMemoryHistory()}>
        <PromoMovieCard onMounted={jest.fn()} {...props}/>
      </Router>
    </redux.Provider>
  ));
};

describe(`PromoMovieCard component`, () => {

  it(`Should render according to the snapshot when there is no movie provided`, () => {
    Object.values(OperationStatus).forEach((movieStatus) => {
      if (movieStatus !== OperationStatus.PENDING) {
        const {container} = renderPromoMovie({movieStatus});
        expect(container).toMatchSnapshot();
      }
    });
  });

  it(`Should render according to the snapshot when promo movie is loading`, () => {
    const {container} = renderPromoMovie({movieStatus: OperationStatus.PENDING});
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot when movie is provided`, () => {
    const movie = {id: `fakeId`};
    const {container} = renderPromoMovie({movie});
    expect(container).toMatchSnapshot();
  });
});
