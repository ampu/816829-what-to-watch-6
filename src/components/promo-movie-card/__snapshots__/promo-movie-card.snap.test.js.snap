// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromoMovieCard component Should render according to the snapshot when movie is provided 1`] = `
<div>
  <section
    class="movie-card"
    style="background-color: black;"
  >
    
    <h1
      class="visually-hidden"
    />
    <header
      class="page-header movie-card__head"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="user-block"
      />
    </header>
    <div
      class="movie-card__wrap"
    >
      <div
        class="movie-card__info"
      >
        
        <div
          class="movie-card__desc"
        >
          <h2
            class="movie-card__title"
          />
          <p
            class="movie-card__meta"
          >
            
            
          </p>
          <div
            class="movie-card__buttons"
          >
            <button
              class="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlink:href="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`PromoMovieCard component Should render according to the snapshot when promo movie is loading 1`] = `
<div>
  <section
    class="movie-card movie-card--empty"
    style="background-color: black;"
  >
    
    <h1
      class="visually-hidden"
    />
    <header
      class="page-header movie-card__head movie-card__head--empty"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="container container--centered container--absolute"
      >
        <div
          class="spinner-loading"
        >
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
      <div
        class="user-block"
      />
    </header>
    <div
      class="movie-card__wrap"
    >
      <div
        class="movie-card__info"
      >
        
      </div>
    </div>
  </section>
</div>
`;

exports[`PromoMovieCard component Should render according to the snapshot when there is no movie provided 1`] = `
<div>
  <section
    class="movie-card movie-card--empty"
    style="background-color: black;"
  >
    
    <h1
      class="visually-hidden"
    />
    <header
      class="page-header movie-card__head movie-card__head--empty"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="user-block"
      />
    </header>
    <div
      class="movie-card__wrap"
    >
      <div
        class="movie-card__info"
      >
        
      </div>
    </div>
  </section>
</div>
`;

exports[`PromoMovieCard component Should render according to the snapshot when there is no movie provided 2`] = `
<div>
  <section
    class="movie-card movie-card--empty"
    style="background-color: black;"
  >
    
    <h1
      class="visually-hidden"
    />
    <header
      class="page-header movie-card__head movie-card__head--empty"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="user-block"
      />
    </header>
    <div
      class="movie-card__wrap"
    >
      <div
        class="movie-card__info"
      >
        
      </div>
    </div>
  </section>
</div>
`;

exports[`PromoMovieCard component Should render according to the snapshot when there is no movie provided 3`] = `
<div>
  <section
    class="movie-card movie-card--empty"
    style="background-color: black;"
  >
    
    <h1
      class="visually-hidden"
    />
    <header
      class="page-header movie-card__head movie-card__head--empty"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="user-block"
      />
    </header>
    <div
      class="movie-card__wrap"
    >
      <div
        class="movie-card__info"
      >
        
      </div>
    </div>
  </section>
</div>
`;
